File: notification_system_impl_go/models/audit_logs.go
package models

type ResourceType string
type ActionType string

const (
	ResourceNotification ResourceType = "notification"
	ResourceUser         ResourceType = "user"
	ResourceWebhook      ResourceType = "webhook"
)

const (
	ActionCreate ActionType = "create"
	ActionUpdate ActionType = "update"
	ActionDelete ActionType = "delete"
)

type AuditLog struct {
	ID           string       `gorm:"type:char(36);primaryKey"`
	UserID       *string      `gorm:"type:char(36);index:idx_audit_user"`
	User         *Users       `gorm:"constraint:OnUpdate:CASCADE,OnDelete:SET NULL;foreignKey:UserID"`
	Action       ActionType   `gorm:"type:varchar(100);not null;index:idx_audit_action"`
	ResourceType ResourceType `gorm:"type:varchar(50);not null;index:idx_audit_resource,priority:1"`
	ResourceID   *string      `gorm:"type:char(36);index:idx_audit_resource,priority:2"`
	IPAddress    *string      `gorm:"type:varchar(45)"`
	UserAgent    *string      `gorm:"type:text"`
	Changes      *string      `gorm:"type:text"`
	CreatedAt    int64        `gorm:"autoCreateTime:milli;index:idx_audit_created"`
	UpdatedAt    int64        `gorm:"autoUpdateTime:milli"`
}

func (AuditLog) TableName() string {
	return "audit_logs"
}

File: notification_system_impl_python/models/audit_logs.py
from configs.db import db
from helpers.helpers import now_ms
import uuid
from helpers.enums import ResourceType, ActionType
from sqlalchemy import Index
from enum import Enum


class AuditLogs(db.Model):
    __tablename__ = "audit_logs"
    __table_args__ = (
        Index("idx_audit_user", "user_id"),
        Index("idx_audit_action", "action"),
        Index("idx_audit_resource", "resource_type", "resource_id"),
        Index("idx_audit_created", "created_at"),
    )

    id = db.Column(db.String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = db.Column(db.String(36), db.ForeignKey("users.id", onupdate="CASCADE", ondelete="SET NULL"), nullable=True)
    user = db.relationship("Users", back_populates="audit_logs", passive_deletes=True)
    action = db.Column(
        db.Enum(
            ActionType,
            values_callable=lambda obj: [e.value for e in obj],
            native_enum=True,
            validate_strings=True,
        ),
        nullable=False,
    )
    resource_type = db.Column(
        db.Enum(
            ResourceType,
            values_callable=lambda obj: [e.value for e in obj],
            native_enum=True,
            validate_strings=True,
        ),
        nullable=False,
    )
    resource_id = db.Column(db.String(36))
    ip_address = db.Column(db.String(45))
    user_agent = db.Column(db.Text)
    changes = db.Column(db.Text)
    createdAt = db.Column("created_at", db.BigInteger, nullable=False, default=now_ms)
    updatedAt = db.Column("updated_at", db.BigInteger, nullable=False, default=now_ms, onupdate=now_ms)

    def __repr__(self):
        return f"<AuditLogs id={self.id}>"

    def to_dict(self):
        data = {}
        for column in self.__table__.columns:
            value = getattr(self, column.name)
            if isinstance(value, Enum):
                value = value.value
            data[column.name] = value
        return data
